{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# AMIA 2016 Annual Symposium Workshop (WG13)\n",
    "\n",
    "### Mining Large-scale Cancer Genomics Data Using Cloud-based Bioinformatics Approaches (RNAseq)\n",
    "\n",
    "Riyue Bao, Ph.D. \n",
    "Center for Research Informatics,\n",
    "The University of Chicago.\n",
    "9:00 AM - 10:15 AM, November 13, 2016\n",
    "\n",
    "***\n",
    "\n",
    "The workshop materials are available on [Github](https://github.com/cribioinfo/CRI-Workshop-AMIA-2016-RNAseq) licensed via [LGPLv3](https://www.gnu.org/licenses/lgpl-3.0.en.html).\n",
    "\n",
    "## Table of Contents: <a class=\"anchor\" id=\"top\"></a>\n",
    "* [Objective](#01)\n",
    "* [Dataset](#02)\n",
    "* [1. RNAseq technology, clinical application and data analysis](#03)\n",
    "\t* [1.1 What is RNAseq and when to use it](#03.1)\n",
    "    * [1.2 Challenges and limitations](#03.2)\n",
    "* [2. How to perform RNAseq analysis](#04)\n",
    "\t* [Workflow](#04.1)\n",
    "\t* [2.1 - 2.2 Quality assessment and preprocessing of raw sequencing reads](#04.2)\n",
    "\t* [2.3 - 2.4 Map reads to reference genome and quantification of transcript abundance](#04.3)\n",
    "\t* [2.5 - 2.7 Normalize gene expression and identify DEGs: DESeq2](#04.4)\n",
    "\t* [2.8 Identify enriched GO and pathways: clusterProfiler](#04.5)\n",
    "* [3. Practice how to associate gene expression with clinical data](#05)\n",
    "\t* [Workflow](#05.1)\n",
    "\t* [3.1 - 3.2. Import data, normalize expression and cluster patients based on expression profile](#05.2)\n",
    "\t* [3.3 Survival analysis](#05.3)\n",
    "* [Summary](#06)\n",
    "* [Run pipeline (after the workshop)](#07)\n",
    "\n",
    "***\n",
    "\n",
    "## Objective<a class=\"anchor\" id=\"01\"></a> <small>[[top](#top)]</small>\n",
    "\n",
    "* **Introduction to RNAseq technology and clinical application** (10 minutes)\n",
    "    * Learn the background and when to use RNAseq for your projects\n",
    "* **How to perform RNAseq analysis** (50 minutes) *hands-on*\n",
    "    * Learn the good-practice analysis protocol (**8 steps**)\n",
    "        * From raw sequencing data to quantitative measures of transcripts (automated pipeline)\n",
    "        * Identify differentially expressed genes (DEGs) between conditions and affected pathways (R & Bioconductor)\n",
    "        * Generate high-quality figures for publication\n",
    "* **Associate gene expression data with clinical outcome** (15 minutes) *hands-on*\n",
    "    * Discover tumor subtypes and detect survival difference (**3 steps**) (R & Bioconductor)\n",
    "\n",
    "<img src='notebook_ext/ipynb_data/assets/Figure24.png', title = 'RNAseq objective', width = 600, height = 600>\n",
    "\n",
    "***\n",
    "\n",
    "## Dataset<a class=\"anchor\" id=\"02\"></a> <small>[[top](#top)]</small>\n",
    "* **Data for RNAseq analysis**\n",
    "    * Two groups (*PRDM11* KO vs WT, human U2932 cells), 6 samples [1]\n",
    "    * Aim to identify DEGs / pathways altered between KO and WT groups\n",
    "\n",
    "| Sample | Group | Sequencing File | Sequencing Data |\n",
    "|------|------|------|------|------|   \n",
    "| KO01 | KO | KO01.fastq.gz | 74,126,025 reads |   \n",
    "| KO02 | KO | KO02.fastq.gz | 64,695,948 reads |   \n",
    "| KO03 | KO | KO03.fastq.gz | 52,972,573 reads |   \n",
    "| WT01 | WT | WT01.fastq.gz | 55,005,729 reads |   \n",
    "| WT01 | WT | WT02.fastq.gz | 61,079,377 reads |   \n",
    "| WT01 | WT | WT03.fastq.gz | 66,517,156 reads |  \n",
    "\n",
    "* **Data for clinical association**\n",
    "    * The Cancer Genome Atlas (TCGA) ovarian cancer, 379 primary tumors **[Genomics Data Commons](https://gdc-portal.nci.nih.gov)** [2]\n",
    "    * Aim to use gene expression to group patients into subtypes and detect survival differences\n",
    "    * **TCGA datasets** (35 cancer types, > 20,000 tumor/normal samples, > 9,000 patients)\n",
    "<img src='notebook_ext/ipynb_data/assets/Figure23.2.png', title = 'GDC', width = 600, height = 600>\n",
    "\n",
    "***\n",
    "\n",
    "## 1. RNAseq technology, clinical application and data analysis<a class=\"anchor\" id=\"03\"></a> <small>[[top](#top)]</small>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Biological and clinical questions\n",
    "\n",
    "* I am interested in studying transcriptional landscape shift before and after drug treatment in cell lines\n",
    "* I want to identify which pathways are affected after knocking down my favorite gene in mice\n",
    "* I have expression data of clinical isolates collected at various time points, when patient's response changed. Why?\n",
    "* I have a cohort of patients and want to discover which gene signature predicts patient's response to treatment\n",
    "* I want to detect gene fusions, expressed mutations, and disrupted isoforms in tumors that may be related to disease    \n",
    "\n",
    "... and more!\n",
    "\n",
    "### 1.1. What is RNAseq and when to use it<a class=\"anchor\" id=\"03.1\"></a> <small>[[top](#top)]</small>\n",
    "<img src='notebook_ext/ipynb_data/assets/Figure14.png', title = 'A typical RNAseq experiment', width = 800, height = 400>\n",
    "* **High-throughput sequencing of RNA: Profile, identify or assemble transcripts**\n",
    "    * **Detect gene expression changes between conditions** (hands-on)\n",
    "    * Identify novel splice sites / exons, mutations, fusion genes, etc.\n",
    "    * Broad detection range (lowly to highly expressed genes), high sensitivity, low requirement of RNA amount\n",
    "    * Available for all species (reference genome is optional): reference genome-guided alignment or *de novo* assembly\n",
    "    * Transcriptome-wide approach for quantitative measurements and gene discovery without prior knowledge\n",
    "* **Experimental Design** ([considerations and limitations](https://github.com/cribioinfo/CRI-Workshop-AMIA-2016-RNAseq/tree/master/notebook_ext/ipynb_data/assets/Figure13.png))\n",
    "    * Include biological replicates for increased discovery power\n",
    "<img src='notebook_ext/ipynb_data/assets/Figure15.2.png', title = 'Biological replicates', width = 400, height = 400>\n",
    "\n",
    "### 1.2 Challenges and limitations<a class=\"anchor\" id=\"03.2\"></a> <small>[[top](#top)]</small>\n",
    "* Relatively poor RNA quality for tumor FFPE samples\n",
    "* Contamination from adjacent normal tissue\n",
    "* Still more expensive than targeted-panel sequencing such as NanoString\n",
    "* 40 million mapped reads are usually sufficient for gene profiling, but > 80 million are required to detect bottom 1% lowly expressed genes [3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "## 2. How to perform RNAseq analysis<a class=\"anchor\" id=\"04\"></a> <small>[[top](#top)]</small>\n",
    "\n",
    "<img src='notebook_ext/ipynb_data/assets/Figure20.2.png', title = 'RNAseq analysis', width = 800, height = 400>\n",
    "\n",
    "### Workflow \n",
    "\n",
    "The best-practice analysis protocol takes 8 major steps. For a more detailed description and program commands used in each step, refer to the extended version of notebooks (`notebook_ext/01.Run_RNAseq.tutorial.ipynb`).\n",
    "\n",
    "* Steps 1-4: From raw sequencing to transcript quantification is **automated through [BigDataScript](https://pcingola.github.io/BigDataScript/) (BDS) pipeline**.\n",
    "* Steps 5-8: DEG and pathway analysis will be practiced **interactively to better interpret the results**.\n",
    "\n",
    "<img src='notebook_ext/ipynb_data/assets/Figure25.png', title = 'RNAseq workflow', width = 900, height = 900>\n",
    "\n",
    "Due to time limit, we will not run the BDS pipeline in this workshop. Since it is automated, participants are encouraged to explore it after the workshop following the instructions at [GitHub](https://github.com/cribioinfo/CRI-Workshop-AMIA-2016-RNAseq/pipeline). The launching of the pipeline only requires submission of one script `Build_RNAseq.sh`, which was already prepared for you. The pipeline was designed to **take care of all the dependencies between tasks/jobs with robust checkpoints and high reproducibility**. \n",
    "\n",
    "### Steps 1-4\n",
    "\n",
    "***\n",
    "\n",
    "### 2.3.1 - 2.3.2 Quality assessment and preprocessing of raw sequencing reads<a class=\"anchor\" id=\"04.2\"></a> <small>[[top](#top)]</small>\n",
    "\n",
    "* **Raw sequencing reads** are stored in FastQ format (e.g. `KO01.fastq.gz`), where each read is presented by 4 lines\n",
    "<img src='notebook_ext/ipynb_data/assets/Figure10.png', title = 'Sequencing reads in FastQ format', width = 600, height = 90>\n",
    "* QC produces reports that help you evaluate if a sequencing run is successful and if reads are of high quality (example in `notebook_ext/01.Run_RNAseq.tutorial.ipynb`, *MultiQC*)\n",
    "* *Optional* Preprocess reads to improve mapping rate and accuracy (in the next step, 1.3.3)\n",
    "    * Trim low-quality bases, clip adapters, etc.\n",
    "    * Avoid overtrimming in RNAseq! ([Williams et al. 2016](http://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-016-0956-2))\n",
    "\n",
    "***\n",
    "### 2.3.3 - 2.3.4 Map reads to reference genome and quantification of transcript abundance<a class=\"anchor\" id=\"04.3\"></a> <small>[[top](#top)]</small>\n",
    "* Read mapping identifies the location in the genome where a sequencing read comes from\n",
    "* Accurate mapping result is the key for quantification and DEG identification\n",
    "* **splice-aware aligner** (e.g. [STAR](https://github.com/alexdobin/STAR))\n",
    "<img src='notebook_ext/ipynb_data/assets/Figure12.png', title = 'RNAseq mapping result', width = 600, height = 400>\n",
    "* **Different aligners may generate very different results** ([Engström et al. 2013](http://www.nature.com/nmeth/journal/v10/n12/full/nmeth.2722.html))\n",
    "\n",
    "***\n",
    "### RNAseq metrics\n",
    "* Collect metrics to evaluate RNA sample quality and identify potential problems (example in `notebook_ext/01.Run_RNAseq.tutorial.ipynb`, *MultiQC*)\n",
    "    * **Is there high-level genomic DNA contamination?**\n",
    "    * **Was ribosome RNA successfully depleted during library prep?**\n",
    "    * Is the RNA highly degraded?\n",
    "    * How do reads distribute on the genome? (exons, introns, intergenic, etc.)\n",
    "    * Does the target gene knocked down in KO samples have reduced expression as expected?   \n",
    "\n",
    "*For detailed answers to those questions, refer to the extended version of notebooks* (`notebook_ext/01.Run_RNAseq.tutorial.ipynb`)\n",
    "\n",
    "<blockquote>\n",
    "Q1: Which sample (S1-4) has the most severe contamination from genomic DNA?   \n",
    "Hint: higher percentage of intergenic reads indicates more severe DNA contamination in RNA samples\n",
    "</blockquote>\n",
    "<blockquote>\n",
    "Q2: Which sample (S1-4) has the least efficient depletion of ribosome RNAs (rRNAs)?   \n",
    "Hint: rRNAs account for > 80% of the whole transcriptome. If not removed, the majority of the sequencing reads will be derived from rRNA and not the coding or lincRNAs.\n",
    "</blockquote>\n",
    "<img src='notebook_ext/ipynb_data/assets/Figure5.png', title = 'Figure5', width = 600, height = 600>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### Steps 5-8\n",
    "\n",
    "* From the pipeline run, we generated read alignment and counts for every annotated gene on the genome\n",
    "* The next step is to utilize the read counts data to detect DEGs\n",
    "* For example, if we visualize *FOS* gene across 6 samples in genome browser (e.g. [IGV](http://software.broadinstitute.org/software/igv/))\n",
    "\n",
    "<img src='notebook_ext/ipynb_data/assets/Figure16.png', title = 'KO vs WT read coverage FOS', width = 600, height = 600>\n",
    "\n",
    "## Hands-on practice START\n",
    "\n",
    "* **<font color='red'>Due to time constraints, and to facilitate workshop demo, small R script snippets were prepared to simplify steps described in hands-on</font>**\n",
    "* Those scripts can be found in directory `notebook_ext/ipynb_data/Rscripts`, e.g. `DESeq2.plot_PCA.R`\n",
    "* Contents of those snippets will not be covered in this workshop but detailed comments were added to every `R` script \n",
    "* Full length description & R code is provided in `notebook_ext/02.Run_RNAseq.tutorial.ipynb`\n",
    "\n",
    "***\n",
    "### 2.3.5 - 2.3.7 Normalize gene expression and identify DEGs: DESeq2<a class=\"anchor\" id=\"04.4\"></a> <small>[[top](#top)]</small>\n",
    "\n",
    "Involves five sub-steps in [R](https://www.r-project.org/about.html) (i to v).\n",
    "\n",
    "**To systematically identify significant DEGs across the transcriptome, we will use [DESeq2](http://bioconductor.org/packages/release/bioc/html/DESeq2.html)**\n",
    "* Different tools (`edgeR`, `limmavoom`) may generate different results! [Seyednasrollah F et al. 2015](https://dx.doi.org/10.1093/bib/bbt086)\n",
    "* We will also practice how to generate publication-quality figures such as PCA, heatmap, clustering, and pathways from our data\n",
    "\n",
    "**Good coding practice**\n",
    "* One statement each line for clarity. It is possible to concatenate statements with `;` but not recommended\n",
    "* Break long statements into multiple lines with less than 80 columns per line for better readability!\n",
    "\n",
    "***\n",
    "### i. Prepare working environment and set up parameters\n",
    "* Set up global variables at the beginning, avoid hard-coding of file names, directories, etc.\n",
    "* Can easily apply to other projects with similar experimental design"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##-- Clean the environment: always run this before starting new analysis!\n",
    "rm(list=ls())\n",
    "\n",
    "##-- Load libraries required in this analysis\n",
    "##-- Those packages were pre-installed in the AWS machine\n",
    "pkg.list <- c('ggplot2', 'RColorBrewer', 'reshape', 'broom',\n",
    "             'DESeq2', 'vsn', 'genefilter', 'pheatmap',\n",
    "             'clusterProfiler', 'pathview', 'IRdisplay')\n",
    "for(pkg in pkg.list) {\n",
    "    suppressMessages(library(pkg, character.only = TRUE))\n",
    "}\n",
    "\n",
    "##-- Parameters\n",
    "cancer <- 'DLBC' ## full name\n",
    "fdr <- 0.05 ## false-discovery rate adjusted p-value\n",
    "fc <- 1.5 ## expression fold of change\n",
    "gene.type <- 'coding' ## protein-coding genes\n",
    "caller <- 'deseq2'\n",
    "group1 <- 'KO' ## knockout\n",
    "group2 <- 'WT' ## wild-type\n",
    "colors <- c('#CC0000', '#00CC00')\n",
    "\n",
    "##-- Set up working directory\n",
    "work.dir <- '.'\n",
    "setwd(work.dir)\n",
    "\n",
    "##-- Input/Output directories\n",
    "in.dir <- 'notebook_ext/ipynb_data/input'\n",
    "out.dir <- 'notebook_ext/ipynb_data/output'\n",
    "\n",
    "##-- Input/Output files\n",
    "expr.file <- paste0(cancer, '.raw_counts.coding.tsv')\n",
    "sample.file <- paste0(cancer, '.sample_group.tsv')\n",
    "\n",
    "##-- Print analysis info\n",
    "script <- file.path('notebook_ext','ipynb_data','Rscripts',\n",
    "                   'DESeq2.print_info.R')\n",
    "source(script)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### ii. Import data\n",
    "* Read counts on coding genes \n",
    "* Sample groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##-- Read data files\n",
    "data.expr <- read.delim(file.path(in.dir, expr.file), header=TRUE, \n",
    "                       stringsAsFactors=FALSE)\n",
    "data.sample <- read.delim(file.path(in.dir, sample.file), header=TRUE, \n",
    "                         stringsAsFactors=FALSE)\n",
    "\n",
    "print('Show the first three lines of gene expression file:')\n",
    "data.expr[1:3,]\n",
    "print('Show the sample group file:')\n",
    "data.sample"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### iii. Principle Component Analysis (PCA)\n",
    "* Both are good assessment to identify mislabeled samples and outliers\n",
    "* All genes are used as input!\n",
    "* PCA - do samples group as expected? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##-- Format data matrix\n",
    "script <- file.path('notebook_ext','ipynb_data','Rscripts',\n",
    "                   'DESeq2.format_data.R')\n",
    "source(script)\n",
    "\n",
    "##-- DESeq2: build DESeqDataSet object, prepare design matrix\n",
    "dds <- DESeqDataSetFromMatrix(countData = as.matrix(data.expr.proc),\n",
    "                             colData = data.sample, \n",
    "                             design = ~ Group)\n",
    "# print(dds)\n",
    "\n",
    "##-- DESeq2: normalize matrix for clustering/heatmap generation\n",
    "vsd <- varianceStabilizingTransformation(dds, blind=FALSE)\n",
    "\n",
    "##-- DESeq2: correlation between bio replicates (skip in workshop)\n",
    "# script <- file.path('notebook_ext','ipynb_data','Rscripts',\n",
    "# 'DESeq2.plot_SMcor_heatmap.R')\n",
    "#source(script)\n",
    "\n",
    "##-- DESeq2: Principal component analysis (PCA) plot\n",
    "script <- file.path('notebook_ext','ipynb_data','Rscripts',\n",
    "                   'DESeq2.plot_PCA.R')\n",
    "source(script)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### iv. Normalize and detect DEGs\n",
    "* Normalize read counts across samples\n",
    "* **<font color='red'>Build design matrices for linear modeling</font>**   \n",
    "\n",
    "| KO | WT |   \n",
    "|------|------|    \n",
    "| 1 | 0 |   \n",
    "| 1 | 0 |   \n",
    "| 1 | 0 |   \n",
    "| 0 | 1 |   \n",
    "| 0 | 1 |   \n",
    "| 0 | 1 |   \n",
    "\n",
    "* Fit data to a linear model\n",
    "* Lowly expressed genes are automatically filtered out (default as mean read counts < 10)\n",
    "* Estimate the **KO vs WT** fold-change and p-values for each gene "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##-- DESeq2: set up variables\n",
    "script <- file.path('notebook_ext','ipynb_data','Rscripts',\n",
    "                   'DESeq2.set_variable.R')\n",
    "source(script)\n",
    "\n",
    "##-- DESeq2: fit the model and identify DEGs\n",
    "print('Running DESeq2 ... ')\n",
    "dds <- DESeq(dds, test=\"Wald\", betaPrior=TRUE)\n",
    "res <- results(dds, \n",
    "              contrast=c(\"Group\",group1,group2), \n",
    "              pAdjustMethod =\"fdr\", \n",
    "              alpha=fdr)\n",
    "print('Done!')\n",
    "\n",
    "##-- DESeq2: peek into DEG data object\n",
    "summary(res)\n",
    "res <- as.data.frame(res)\n",
    "\n",
    "##-- DESeq2: add anti-log2 fold change and gene symbol to output\n",
    "script <- file.path('notebook_ext','ipynb_data','Rscripts',\n",
    "                   'DESeq2.add_fc_symbol.R')\n",
    "source(script)\n",
    "\n",
    "print('Show the first three lines of DESeq2 output:')\n",
    "res.print[1:3,]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### v. Filter for significant DEGs and plot heatmap\n",
    "* Filter for significant DEGs with at least 1.5 fold-change and FDR-adjusted p-value < 0.05\n",
    "* Generate expression heatmap on significant DEGs (different from PCA and correlation heatmap described in iii)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##-- DESeq2: filter significant DEGs by fc 1.5 and adjusted p < 0.05\n",
    "print('Show the first three lines of DESeq2 output:')\n",
    "res.print[1:3,]\n",
    "script <- file.path('notebook_ext','ipynb_data','Rscripts',\n",
    "                   'DESeq2.flt_sigDEGs.R')\n",
    "source(script)\n",
    "\n",
    "##-- DESeq2: plot sig DEG heatmap\n",
    "print(paste0('Heatmap = ', nrow(data.plot), ' genes on the row, ', \n",
    "             ncol(data.plot), ' samples on the column'))\n",
    "pheatmap(data.plot,\n",
    "         scale = 'row',\n",
    "         cluster_rows = TRUE, cluster_cols = TRUE,\n",
    "         show_rownames = FALSE, show_colnames = TRUE,\n",
    "         clustering_distance_rows = \"euclidean\",\n",
    "         clustering_distance_cols = \"euclidean\", \n",
    "         clustering_method = \"complete\",\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 2.3.8 Identify enriched GO and pathways: clusterProfiler<a class=\"anchor\" id=\"04.5\"></a> <small>[[top](#top)]</small>\n",
    "* Gene Ontology bological processes and KEGG pathways\n",
    "* Enrichment in significant DEGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##-- clusterProfiler: NOTE those commands were tested for clusterProfiler \n",
    "##-- version 3.0.5 or above. Older versions may not work as they have \n",
    "##-- options revised/removed in the new versions.\n",
    "\n",
    "##-- clusterProfiler: prepare input for GO/KEGG enrichment analysis\n",
    "script <- file.path('notebook_ext','ipynb_data','Rscripts',\n",
    "                   'clusterProfilter.prepare_input.R')\n",
    "source(script)\n",
    "\n",
    "##-- clusterProfiler: GO over-representation test\n",
    "##-- takes ~2 minutes .. skip in workshop!\n",
    "# ego <- enrichGO( gene          = genes.sig.anno$ENTREZID,\n",
    "#                 universe      = genes.all.anno$ENTREZID,\n",
    "#                 OrgDb         = 'org.Hs.eg.db',\n",
    "#                 ont           = \"BP\",\n",
    "#                 pAdjustMethod = \"fdr\",\n",
    "#                 pvalueCutoff  = 0.05,\n",
    "#                 readable      = TRUE)\n",
    "##-- instead, load already generated output\n",
    "load(paste0(out.prefix, '.flt.fdr', fdr,'_fc',fc, '.enrichGO.RData'))\n",
    "summary(ego)[1:3,1:7]\n",
    "\n",
    "##-- clusterProfiler: KEGG over-representation test\n",
    "##-- takes ~2 minutes .. skip in workshop!\n",
    "# kk <- enrichKEGG(gene          = genes.sig.anno$ENTREZID,\n",
    "#                  universe      = genes.all.anno$ENTREZID,\n",
    "#                  organism      = \"human\",\n",
    "#                  pAdjustMethod = \"fdr\",\n",
    "#                  pvalueCutoff  = 0.05)\n",
    "##-- instead, load already generated output\n",
    "load(paste0(out.prefix, '.flt.fdr', fdr,'_fc',fc,'.enrichKEGG.RData'))\n",
    "summary(kk)[1:3,]\n",
    "\n",
    "##-- clusterProfiler: visualize pathway with all genes\n",
    "gene.list <- genes.all.anno$foldChange\n",
    "names(gene.list) <- genes.all.anno$ENTREZID\n",
    "pathway <- 'hsa04010'\n",
    "plot.title <- paste0(pathway, '.gene_all.pathview.png')\n",
    "p4 <- pathview(gene.data  = gene.list,\n",
    "              pathway.id = pathway,\n",
    "              species    = \"human\",\n",
    "              limit      = list(gene=4, cpd=1),\n",
    "              kegg.dir   = paste0(out.dir,'/',caller),\n",
    "              out.suffix = paste0('gene_all.pathview'))\n",
    "\n",
    "##-- clusterProfiler: visualization of GO and KEGG enrichment results\n",
    "barplot(ego, showCategory=20)\n",
    "dotplot(ego, showCategory=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##-- KEGG pathwayview written into local files. Display the plot in R\n",
    "display_png(file='notebook_ext/ipynb_data/assets/hsa04010.gene_all.pathview.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "## 3. Practice how to associate gene expression with clinical data<a class=\"anchor\" id=\"05\"></a> <small>[[top](#top)]</small>  \n",
    "\n",
    "* TCGA hosted on GDC (https://gdc-portal.nci.nih.gov)\n",
    "<img src='notebook_ext/ipynb_data/assets/Figure19.png', title = 'GDC website', width = 800, height = 200>\n",
    "* **Our data**\n",
    "    * ~600 Ovarian Cancer patients (expression and clinical)\n",
    "<img src='notebook_ext/ipynb_data/assets/Figure18.png', title = 'TCGA OV Nature paper', width = 800, height = 200>\n",
    "\n",
    "### Workflow<a class=\"anchor\" id=\"05.1\"></a> <small>[[top](#top)]</small>  \n",
    "\n",
    "The analysis takes 3 major steps. Full-length description & R code is provided in `notebook_ext/03.Run_RNAseq.tutorial.ipynb`.\n",
    "\n",
    "<img src='notebook_ext/ipynb_data/assets/Figure26.png', title = 'TCGA & Clinical workflow', width = 900, height = 900>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.1 - 3.2. Import data, normalize expression and cluster patients based on expression profile<a class=\"anchor\" id=\"05.2\"></a> <small>[[top](#top)]</small>\n",
    "\n",
    "* Select 150 most variable genes for sample clustering\n",
    "    * Without patient groups known, using all genes for sample clustering is computational intensive and may mask the true differences between potential sample groups\n",
    "    * The most variable genes likely capture the majority of the sample variation   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##-- Clean up environment\n",
    "rm(list=ls())\n",
    "set.seed(13)\n",
    "\n",
    "##-- Load libraries required in this analysis\n",
    "##-- Those packages were pre-installed in the AWS machine\n",
    "pkg.list <- c('ggplot2', 'RColorBrewer', 'reshape', 'broom',\n",
    "             'gplots', 'survival', 'NMF', 'IRdisplay')\n",
    "for(pkg in pkg.list) {\n",
    "    suppressMessages(library(pkg, character.only = TRUE))\n",
    "}\n",
    "\n",
    "##-- Parameters\n",
    "cancer <- 'OV' \n",
    "gene.top.count <- 150\n",
    "\n",
    "##-- Import data\n",
    "script <- file.path('notebook_ext','ipynb_data','Rscripts',\n",
    "                   'TCGA_OV.import_data.R')\n",
    "source(script)\n",
    "\n",
    "##-- run NMF to cluster samples & genes \n",
    "##-- (use 4 core 'p4', and print more info 'v')\n",
    "##-- takes 15 minutes to run... skip in workshop\n",
    "print('NMF clustering start...')\n",
    "# expr.sub.nmf <- nmf(expr.sub, \n",
    "#                    rank = 3, \n",
    "#                    method = 'brunet', \n",
    "#                    seed = 'random', \n",
    "#                    nrun = 100, \n",
    "#                    .opt = 'vp4') \n",
    "##-- workshop only: load already generated result\n",
    "load(paste0(out.dir,'/',expr.file,'.nmf.RData'))\n",
    "print('Done!')\n",
    "\n",
    "##-- plot expression heatmap\n",
    "script <- file.path('notebook_ext','ipynb_data','Rscripts',\n",
    "                   'TCGA_OV.plot_heatmap.R')\n",
    "source(script)\n",
    "\n",
    "# print(table(expr.sub.nmf.smclr$cluster))\n",
    "# print(table(expr.sub.nmf.geneclr$cluster))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### 3.2 Survival analysis<a class=\"anchor\" id=\"05.2\"></a> <small>[[top](#top)]</small>\n",
    "\n",
    "Involves two sub-steps in [R](https://www.r-project.org/about.html) (i to ii).\n",
    "\n",
    "* Identify overall survival (OS) difference between the three tumor subtypes / patient groups\n",
    "* Kaplan–Meier (KM) plot, log-rank test, and cox proportional hazards model (univariate and multivariate)   \n",
    "\n",
    "### i. KM plot and log-rank test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##-- Set up R plot display options in notebook\n",
    "options(jupyter.plot_mimetypes = \"image/svg+xml\") \n",
    "options(repr.plot.width = 5, repr.plot.height = 5)\n",
    "\n",
    "##-- add right censoring\n",
    "table(clinical$vital.status)\n",
    "clinical$censor <- NA\n",
    "clinical$censor[which(clinical$vital.status == 'LIVING')] <- 0\n",
    "clinical$censor[which(clinical$vital.status == 'DECEASED')] <- 1\n",
    "clinical$censor <- as.numeric(clinical$censor)\n",
    "\n",
    "##-- KM plot \n",
    "surv <- Surv(clinical$overall.survival.day, clinical$censor)\n",
    "surv.fit <- survfit(surv ~ clinical$cluster)\n",
    "plot(surv.fit, mark=4, col=c('#00CC00','#0000CC','#CC0000'), \n",
    "     lty=1, lwd=1.5,cex=0.8,cex.lab=1.5, cex.axis=1.5, cex.main=1,\n",
    "     main='Kaplan-Meier survival curves for TCGA OV dataset',\n",
    "     xlab='Overall Survival (days)', \n",
    "     ylab='Probability of Survival')\n",
    "text(1500,0.08,  labels=paste0('cluster1 (n=',sample.counts[1,2],')'), \n",
    "     cex=1.2, col='#00CC00')\n",
    "text(650,0.40, labels=paste0('cluster2\\n(n=',sample.counts[2,2],')'), \n",
    "     cex=1.2, col='#0000CC')\n",
    "text(2500,0.68, labels=paste0('cluster3 (n=',sample.counts[3,2],')'), \n",
    "     cex=1.2, col='#CC0000')\n",
    "\n",
    "separator\n",
    "\n",
    "##-- log-rank test\n",
    "print(paste0('Running log-rank test for survival'))\n",
    "surv <- Surv(clinical$overall.survival.day, clinical$censor)\n",
    "surv.diff <- survdiff(surv ~ clinical$cluster)\n",
    "surv.diff\n",
    "\n",
    "separator"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The KM plot is similar to TCGA OV paper Figure 2e.\n",
    "** <font color='red'>Note that patient cluster C1 in the TCGA OV nature paper corresponds to C3 in our analysis!</font>**\n",
    "\n",
    "<img src='notebook_ext/ipynb_data/assets/Figure27.png', title = 'TCGA OV Nature paper Fig 2e', width = 300, height = 300>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "### ii. Cox multivariate and univariate model\n",
    "* Include one factor of interest (e.g. cluster 1/2/3): univariate\n",
    "* Include multiple factors such as cluster, age, tumor grade, and tumor stage: multivariate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "##-- cox pp haz model (comparing cluster 2 and 3 as an example!)\n",
    "print(paste0('Running Cox proportional hazards model for survival'))\n",
    "clinical.sub <- clinical[clinical$cluster %in% c(1,2,3),]\n",
    "surv <- Surv(clinical.sub$overall.survival.day, \n",
    "            clinical.sub$censor)\n",
    "\n",
    "print('Cluster 1 will be set as baseline ... ')\n",
    "\n",
    "separator\n",
    "\n",
    "##-- univariate cox model \n",
    "print(paste0('(a) Univariate cox model for survival'))\n",
    "m1 <- coxph(surv ~ (clinical.sub$cluster))\n",
    "tidy(m1)\n",
    "\n",
    "separator\n",
    "\n",
    "##-- full cox model\n",
    "print(paste0('(b) Full multivariate cox model for survival'))\n",
    "m2 <- coxph(surv ~ (clinical.sub$cluster + \n",
    "                    clinical.sub$age.at.diagnosis.year +\n",
    "                    clinical.sub$tumor.grade))\n",
    "tidy(m2)\n",
    "\n",
    "separator\n",
    "\n",
    "##-- reduced cox model\n",
    "##-- Always use the simplest model with the least necessary amount \n",
    "##-- of covariates to minimize overfitting\n",
    "print(paste0('(c) Reduced multivariate cox model for survival'))\n",
    "m3 <- coxph(surv ~ (clinical.sub$cluster +\n",
    "                    clinical.sub$age.at.diagnosis.year))\n",
    "tidy(m3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Hands-on practice END"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "print('Program finished!')\n",
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "## Summary<a class=\"anchor\" id=\"06\"></a> <small>[[top](#top)]</small>\n",
    "\n",
    "In this workshop, we demonstrated how to run RNAseq analysis to identify DEGs & pathways, and how to use expression data to identify patient groups with different clinical outcomes. We introduced the commonly used bioinformatics tools and good-practice approaches. All the analysis was performed on a pre-built machine in AWS EC2 cloud. \n",
    "\n",
    "All the class materials will stay open on [GitHub](https://github.com/cribioinfo/CRI-Workshop-AMIA-2016-RNAseq) if you want to practice or use the pipelines and scripts for your own research projects.\n",
    "\n",
    "Questions? Post on Github or contact Riyue (Sunny) at `rbao AT bsd dot uchicago dot edu`.\n",
    "\n",
    "<img src='notebook_ext/ipynb_data/assets/Figure21.png', title = 'RNAseq summary', width = 600, height = 600>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "\n",
    "## Run pipeline (after the workshop)<a class=\"anchor\" id=\"07\"></a> <small>[[top](#top)]</small>\n",
    "\n",
    "### 1. Open terminal from Jupyter Notebook\n",
    "\n",
    "Go to your Jupyter Notebook homepage. Click [New] button at top right corner of the homepage. In the dropdown menu, click [Terminal]. \n",
    "\n",
    "### 2. Get familiar with the file structure\n",
    "\n",
    "All workshop materials are openly accessible via [GitHub](https://github.com/cribioinfo/CRI-Workshop-AMIA-2016-RNAseq). Documentation and tutorials of the pipelines can be found in [README](https://github.com/cribioinfo/CRI-Workshop-AMIA-2016-RNAseq/tree/master/pipeline) and [wiki](https://github.com/cribioinfo/CRI-Workshop-AMIA-2016-RNAseq/wiki).\n",
    "\n",
    "* **Run_RNAseq.tutorial.ipynb**: The main lecture / hands-on notebook (covered in workshop)\n",
    "* **Run_RNAseq.tutorial.rendered.ipynb**: Same as above, but with all outputs & figures rendered\n",
    "* **pipeline** directory: the automated, scalable and modularized pipeline \n",
    "    * From raw sequencing reads to quantified transcript abundance through one click\n",
    "    * [`BigDataScript`](https://pcingola.github.io/BigDataScript/) & `Perl`\n",
    "* **notebook_ext** directory: Extended notebooks 01-03 (not covered in workshop)\n",
    "    * lecture notes\n",
    "    * R codes and comments \n",
    "\n",
    "`##-- commands`   \n",
    "`pwd`   \n",
    "`cd ~/CRI-Workshop-AMIA-2016-RNAseq`   \n",
    "`ll pipeline`   \n",
    "`ll notebook_ext`   \n",
    "\n",
    "### 2. Launch pipeline (takes ~ 5 minutes)\n",
    "\n",
    "`##-- commands `   \n",
    "`pwd`   \n",
    "`cd ~/CRI-Workshop-AMIA-2016-RNAseq/pipeline`      \n",
    "`cd test`   \n",
    "`./Build_RNAseq.DLBC.sh`   \n",
    "`##-- START Thu Oct 27 15:57:38 UTC 2016  Running ../Build_RNAseq.pl`   \n",
    "`##-- START Thu Oct 27 15:57:38 UTC 2016  Running Submit_RNAseq.DLBC.sh`   \n",
    "`##-- running ... 3 ~ 4 minutes`   \n",
    "`##-- END Thu Oct 27 16:01:25 UTC 2016`      \n",
    "\n",
    "***\n",
    "\n",
    "## Next Session: ChIPseq analysis and Integration of RNAseq & ChIPseq (Kyle Hernandez PhD)\n",
    "\n",
    "***\n",
    "\n",
    "## References\n",
    "[1] Fog. et al. 2015. [Loss of PRDM11 promotes MYC-driven lymphomagenesis](http://www.bloodjournal.org/content/125/8/1272.long?sso-checked=true). Blood 125(8):1272-81   \n",
    "[2] The Cancer Genome Atlas Research Network. 2011. [Integrated genomic analyses of ovarian carcinoma](http://www.nature.com/nature/journal/v474/n7353/full/nature10166.html). Nature 474, 609–615)   \n",
    "[3] Sims et al. 2014. [Sequencing depth and coverage: key considerations in genomic analyses](http://www.nature.com/nrg/journal/v15/n2/abs/nrg3642.html). Nature Reviews Genetics 15, 121–132\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
