{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# AMIA 2016 Annual Symposium Workshop (WG13)\n",
    "\n",
    "### Mining Large-scale Cancer Genomics Data Using Cloud-based Bioinformatics Approaches (RNAseq)\n",
    "### * Part 01 RNAseq technology, clinical application and data analysis\n",
    "\n",
    "Riyue Bao, Ph.D. \n",
    "Center for Research Informatics,\n",
    "The University of Chicago.\n",
    "November 13, 2016\n",
    "\n",
    "***\n",
    "\n",
    "The workshop materials are accessible on [Github](https://github.com/cribioinfo/CRI-Workshop-AMIA-2016-RNAseq).\n",
    "\n",
    "## Table of Contents: <a class=\"anchor\" id=\"top\"></a>\n",
    "* [Objective](#01)\n",
    "* [Workflow](#02)\n",
    "* [Dataset](#03)\n",
    "* [Pipeline](#04)\n",
    "* [Run pipeline](#05)\n",
    "* [How to perform RNAseq analysis](#06)\n",
    "\t* [1. Quality assessment of raw sequencing reads: FastQC](#06.1)\n",
    "\t* [2. Preprocessing: Trimmomatic](#06.2)\n",
    "\t* [3. Map reads to reference genome (GRCh38): STAR](#06.3)\n",
    "        * [Collect RNAseq quality metrics & coverage: Picardtools, bedtools, RSeQC](#06.10)\n",
    "        * [How to use those metrics to decide if a sample has potential quality problem?](#06.11)\n",
    "\t* [4. Quantify transcript abundance: featureCounts](#06.4)\n",
    "        * [BDS pipeline run report](#06.12)\n",
    "\t* [5-7. Identify differentially expressed genes (DEGs) between conditions: DESeq2](#06.5)\n",
    "\t* [8. Identify biological processes and pathways enriched in genes of interest: clusterProfilter](#06.8)\n",
    "\n",
    "***\n",
    "\n",
    "## Objective<a class=\"anchor\" id=\"01\"></a> <small>[[top](#top)]</small>\n",
    "\n",
    "* Learn the good-practice RNAseq analysis pipeline\n",
    "* Learn commonly used bioinformatics tools\n",
    "* Practice the automated, scalable pipeline\n",
    "* Explore the quality metrics and input/output of the RNAseq pipeline\n",
    "* Visualize result files and quality plots\n",
    "\n",
    "***\n",
    "\n",
    "## Workflow<a class=\"anchor\" id=\"02\"></a> <small>[[top](#top)]</small>\n",
    "\n",
    "<img src='ipynb_data/assets/RNAseq.workflow.png', title = 'RNAseq workflow', width = 1000, height = 1000>\n",
    "\n",
    "***\n",
    "\n",
    "## Dataset<a class=\"anchor\" id=\"03\"></a> <small>[[top](#top)]</small>\n",
    "\n",
    "The test datasets used in this workshop are from \n",
    "Fog. et al. 2015. Loss of PRDM11 promotes MYC-driven lymphomagenesis. Blood, 125(8):1272-81\n",
    "<http://www.bloodjournal.org/content/125/8/1272.long?sso-checked=true>\n",
    "\n",
    "***\n",
    "\n",
    "## Pipeline<a class=\"anchor\" id=\"04\"></a> <small>[[top](#top)]</small>\n",
    "\n",
    "For this workshop, the machine you are using has everything pre-compiled and pre-installed. It is ready for analysis.\n",
    "\n",
    "In the future, if you'd like to use the pipleine on your own machines, download analysis pipeline from [Github](https://github.com/cribioinfo/CRI-Workshop-AMIA-2016-RNAseq) and follow the instructions to install.\n",
    "\n",
    "`git clonehttps://github.com/cribioinfo/CRI-Workshop-AMIA-2016-RNAseq.git`     \n",
    "\n",
    "Detailed documentation of the pipeline can be found on Github [README](https://github.com/cribioinfo/CRI-Workshop-AMIA-2016-RNAseq) and [wiki](https://github.com/cribioinfo/CRI-Workshop-AMIA-2016-RNAseq/wiki).\n",
    "\n",
    "***\n",
    "\n",
    "## Run pipeline<a class=\"anchor\" id=\"05\"></a> <small>[[top](#top)]</small>\n",
    "\n",
    "### 1. Open terminal from Jupyter Notebook\n",
    "\n",
    "At Jupyter homepage, go to [New] button at the top-right corner of the notebook. In the dropdown menu, click [Terminal]. \n",
    "\n",
    "### 2. Get familiar with the file structure\n",
    "\n",
    "This repository contains the following items:\n",
    "* Run_RNAseq.tutorial.ipynb - the main notebook for lecture and hands-on\n",
    "* Run_RNAseq.tutorial.rendered.ipynb - same as above, but with all outputs & figures already rendered for browsing\n",
    "* notebook_ext/ directory: the extended version of contents covered in the main notebook\n",
    "* **pipeline/** directory: the automated, scalable and modularized pipeline \n",
    "    * From raw sequencing reads to quantified transcript abundance through one click\n",
    "    * [`BigDataScript`](https://pcingola.github.io/BigDataScript/) & `Perl`\n",
    "      \n",
    "`##-- commands`       \n",
    "`pwd`    \n",
    "`cd ~/CRI-Workshop-AMIA-2016-RNAseq`    \n",
    "`ls -al`    \n",
    "`ls -al notebook_ext`    \n",
    "`ls -al pipeline`    \n",
    "\n",
    "### 3. Launch pipeline (takes ~ 5 minutes)\n",
    "     \n",
    "`##-- commands`    \n",
    "`pwd`    \n",
    "`cd ~/CRI-Workshop-AMIA-2016-RNAseq/pipeline/test`    \n",
    "`./Build_RNAseq.DLBC.sh`    \n",
    "`##-- START Thu Oct 27 15:57:38 UTC 2016  Running ../Build_RNAseq.pl`    \n",
    "`##-- START Thu Oct 27 15:57:38 UTC 2016  Running Submit_RNAseq.DLBC.sh`    \n",
    "`##-- running ... 3 ~ 4 minutes`    \n",
    "`##-- END Thu Oct 27 16:01:25 UTC 2016`    \n",
    "\n",
    "***\n",
    "\n",
    "## How to perform RNAseq analysis<a class=\"anchor\" id=\"06\"></a> <small>[[top](#top)]</small>\n",
    "\n",
    "* Steps 1 - 4 : Automated pipeline (Run_RNAseq.bds)\n",
    "    * Pipeline results running on the complete dataset (for visualization and DEG analysis): \n",
    "        `/home/ubuntu/data/rnaseq/fullset`\n",
    "    * Pipeline results running on the test dataset (for testing only): \n",
    "        `/home/ubuntu/dev/rnaseq/subset`\n",
    "* Steps 5 - 8 : Interactive R & Bioconductor (Notebook 02)\n",
    "\n",
    "<blockquote>\n",
    "Unless pointed out otherwise, all commands shown apply to the test datasets only. Do not use them directly on other datasets. Refer to the pipeline documentation for instructions on how to set up the pipeline for your own projects.\n",
    "</blockquote>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Steps 1 - 4 are run by automated pipeline\n",
    "\n",
    "### 1. Quality assessment of raw sequencing reads: FastQC<a class=\"anchor\" id=\"06.1\"></a> <small>[[top](#top)]</small>\n",
    "\n",
    "* **Goal**\n",
    "    * Check if the reads are of high quality\n",
    "    * Check if any preprocessing step is required (e.g. base trimming, adapter clipping, read filtering)\n",
    "* **Method**\n",
    "    * [FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/), *version 0.11.5*\n",
    "    * Scan raw sequencing reads and produce QC reports for evaluation\n",
    "* **Our data**\n",
    "    * Read quality pretty good (baseQ >= 30 in all base positions)\n",
    "    * Preprocessing is optional\n",
    "    * In this workshop, we will still run the step for demonstration\n",
    "\n",
    "<blockquote>\n",
    "```\n",
    "fastqc --extract -o $out.dir -t 2 --nogroup $r1.fastq.gz\n",
    "```\n",
    "</blockquote>\n",
    "\n",
    "<img src='ipynb_data/assets/Figure1.png', title = 'Figure1', width = 600, height = 600>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. (optional) Preprocessing: Trimmomatic<a class=\"anchor\" id=\"06.2\"></a> <small>[[top](#top)]</small>\n",
    "\n",
    "* **Goal**\n",
    "    * Clean up reads for improved alignment rate and accuracy (for the next step)\n",
    "* **Method**\n",
    "    * [Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic), *version 0.36*\n",
    "    * **Clip adapters**\n",
    "    * **Trim leading/trailing low quality or N bases**\n",
    "    * Trim reads to a specific length\n",
    "    * Filter out reads of low average quality / of specific length\n",
    "    * Convert base quality scores between Phred33 and Phred64 FastQ format\n",
    "* **Avoid over trimming in RNAseq!**\n",
    "    * Williams et al. 2016. Trimming of sequence reads alters RNA-Seq gene expression estimates. BMC Bioinformatics. [DOI: 10.1186/s12859-016-0956-2](http://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-016-0956-2)\n",
    "* **Our data** (KO01 as an example)\n",
    "    * 74,126,025 reads total. Survived: 73,636,793  (99.34%) Dropped: 489,232 (0.66%)\n",
    "\n",
    "<blockquote>\n",
    "```\n",
    "java -Xmx4G -jar trimmomatic-0.36.jar SE -threads 4 -phred33 $r1.fastq.gz $r1.trim.fastq.gz ILLUMINACLIP:TruSeq3-SE.fa:2:30:10 LEADING:5 TRAILING:5 MINLEN:36 SLIDINGWINDOW:4:15\n",
    "```\n",
    "</blockquote>\n",
    "\n",
    "<img src='ipynb_data/assets/Figure2.png', title = 'Figure2', width = 600, height = 600>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Map reads to reference genome (GRCh38): STAR<a class=\"anchor\" id=\"06.3\"></a> <small>[[top](#top)]</small>\n",
    "\n",
    "* **Goal**\n",
    "    * Identify the location in the genome where a sequencing read comes from\n",
    "    * A read may map to one, or multiple genomic locations\n",
    "    * For each read, alignment of more than one target regions is ranked by alignment score\n",
    "    * Accurate mapping result is the key for DEG identification\n",
    "* **Method**\n",
    "    * [STAR](http://bioinformatics.oxfordjournals.org/content/early/2012/10/25/bioinformatics.bts635), *version 2.5.2b*\n",
    "    * **splice-aware aligner**\n",
    "    * Ultrafast (~15 minutes for 50m reads), require lots of memory (~36GB for human genome)\n",
    "    * Flexible options to allow canonical/non-canonical junctions, with/wo known gene annotations, etc.\n",
    "* **Different aligners may generate very different results**\n",
    "    * Engstr√∂m et al. 2013. Systematic evaluation of spliced alignment programs for RNA-seq data. Nat Methods. [DOI: 10.1038/nmeth.2722](http://www.nature.com/nmeth/journal/v10/n12/full/nmeth.2722.html)\n",
    "* **Our data**\n",
    "    * *PRDM11* knockdown U2932 cells in triplicates (KO01-03 vs WT01-03)\n",
    "    * **NOT the full *PRDM11* gene is knockdown!**\n",
    "    * siRNAs target exons 4,5,6 of *PRDM11*, thus only those three exons show a reduction of expression in the KO samples; other exons are not affected\n",
    "    * *PRDM11* knockdown leads to upregulation of *FOS* expression \n",
    "\n",
    "<blockquote>\n",
    "```\n",
    "STAR --runMode alignReads --genomeLoad NoSharedMemory --outFileNamePrefix $out.prefix --readFilesCommand zcat --genomeDir $refgenome.dir --readFilesIn $r1.trim.fastq.gz --runThreadN 2 --outSAMstrandField intronMotif --outFilterIntronMotifs RemoveNoncanonicalUnannotated --outSAMtype BAM SortedByCoordinate \n",
    "```\n",
    "</blockquote>\n",
    "\n",
    "<img src='ipynb_data/assets/Figure3.png', title = 'Figure3', width = 750, height = 1000>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Collect RNAseq quality metrics & coverage: Picardtools, bedtools, RSeQC<a class=\"anchor\" id=\"06.10\"></a> <small>[[top](#top)]</small>\n",
    "\n",
    "* **Goal**\n",
    "    * Evaluate the quality of reads and alignments\n",
    "    * Identify potential problems regarding (RNA) sample quality\n",
    "        * Is the RNA highly degraded?\n",
    "        * Is there high-level genomic DNA contamination?\n",
    "        * Was ribosome RNA sucessfully depleted?\n",
    "        * How do reads distribute on the genome? (exons, introns, intergenic, etc.)\n",
    "        * Is the strandness of read alignment consistent with library type? \n",
    "            * non-stranded or forward/reverse strand-specific libs\n",
    "        * Does the target gene that was knocked down in KO samples have reduced expression, compared to WT?\n",
    "        \n",
    "* **Method**\n",
    "    * [Picardtools](https://broadinstitute.github.io/picard/), *version 2.6.0*\n",
    "        * Utilities *CollectMultipleMetrics* *CollectRnaSeqMetrics*\n",
    "        * Read mapping rate, number of correctly mapped reads, mismatch%\n",
    "        * Distribution of reads mapped to genomic features (e.g. exon, intron, intergenic, CD, UTR, etc.)\n",
    "        * Percentage of reads mapped to ribosome RNAs\n",
    "        * Gene body coverage, read clipping profile, etc.\n",
    "    * [RSeQC](http://rseqc.sourceforge.net/), *version 2.6.4*\n",
    "        * Strandness of mapped reads, etc. \n",
    "    * [bedtools](http://bedtools.readthedocs.io/en/latest/), *version 2.26.0*\n",
    "        * Utilities *genomecov*\n",
    "        * Calculate read coverage on the genome at per-base resolution\n",
    "        * Scaling factor can be supplied to normalize coverage across samples (library size)    \n",
    "        <blockquote>\n",
    "        ```\n",
    "        java -Xmx4G -jar picard.jar CollectRnaSeqMetrics I=$sample.bam O=$out.rnaseq_metrics REF_FLAT=$refgeneanno.refflat.txt  RIBOSOMAL_INTERVALS=$refgeneanno.rRNA.interval_list STRAND=$strandness CHART=$out.rnaseq_metrics.pdf METRIC_ACCUMULATION_LEVEL=SAMPLE VALIDATION_STRINGENCY=LENIENT\n",
    "        java -Xmx4G -jar picard.jar CollectMultipleMetrics I=$sample.bam O=$out R=$refgenome.fa PROGRAM=CollectAlignmentSummaryMetrics PROGRAM=CollectInsertSizeMetrics PROGRAM=QualityScoreDistribution PROGRAM=MeanQualityByCycle VALIDATION_STRINGENCY=LENIENT\n",
    "        ```\n",
    "        </blockquote>\n",
    "        \n",
    "        <blockquote>\n",
    "        ```\n",
    "        infer_experiment.py -i $sample.bam -r $refgeneanno.bed12 -s 200000 > $sample.infer_experiment\n",
    "        ```\n",
    "        </blockquote>        \n",
    "   \n",
    "        <blockquote>\n",
    "        ```\n",
    "        bedtools genomecov -bga -split -ibam $sample.bam -scale $scale.factor > $sample.cov.bedgraph \n",
    "        bedGraphToBigWig $sample.cov.bedgraph $refgenome.chrom.size $sample.cov.bigwig\n",
    "        ```\n",
    "        </blockquote>  \n",
    "* **Combine all metrics into one report!**\n",
    "    * [MultiQC](http://multiqc.info/), *version 0.8*\n",
    "    * 26 supported tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"1000\"\n",
       "            height=\"700\"\n",
       "            src=\"ipynb_data/assets/multiqc_report.html\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7f2ee82dd610>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import IFrame\n",
    "IFrame('ipynb_data/assets/multiqc_report.html', width=1000, height=700)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  **How to use those metrics to decide if a sample has potential quality problem?**<a class=\"anchor\" id=\"06.11\"></a> <small>[[top](#top)]</small>\n",
    "\n",
    "* **Our data** (left panel); Other (right panel)\n",
    "\n",
    "#### 4.1 RNA degradation\n",
    "* RNA quality: Fresh samples (e.g. cell line) > frozen samples (e.g. mouse tissues) >> FFPE samples (e.g. human tumors)\n",
    "* During lib prep, RNA quality is inferred by The RNA integrity number (RIN) evaluated using the 28S to 18S rRNA ratio\n",
    "* In most cases, RIN > 4.5 is recommended\n",
    "* However, studies have shown that RIN can be quite inaccurate for FFPE samples\n",
    "* A better accessment is **Gene body coverage plot** (Picardtools). FFPE samples often has 5' degradation\n",
    "\n",
    "<img src='ipynb_data/assets/Figure4.png', title = 'Figure4', width = 600, height = 600>\n",
    "\n",
    "#### 4.2 Genomic DNA contamination\n",
    "* In lib prep, DNase digestion removes genomic DNA\n",
    "* While fresh samples often has good-quality RNAs, highly-degraded samples (e.g. FFPE tumors) often has a higher degree of genomic DNA contamination\n",
    "* Sometimes DNA contamination could occupy 70% of sequencing reads, greately reducing the discovery power of DEG analysis\n",
    "* Good accessment to identify and estimate genomic DNA contamination includes\n",
    "    * Read distribution in genomic features: **high fraction of intergenic reads indicates DNA contamination**\n",
    "    * Visualize integernic region in genome browser (e.g. [IGV](http://software.broadinstitute.org/software/igv/))\n",
    "\n",
    "<blockquote>\n",
    "Which sample (S1-4) has the most severe contamination from genomic DNA?\n",
    "</blockquote>\n",
    "\n",
    "<img src='ipynb_data/assets/Figure5.png', title = 'Figure5', width = 600, height = 600>\n",
    "\n",
    "#### 4.3 Ribosome RNA fraction\n",
    "* rRNA accounts for > 80% of the transcriptome\n",
    "* Mmost RNAseq lib prep protocol includes an ribosomal RNA depletion step \n",
    "* However, if RNA quality is relatively poor (e.g. FFPE tumors), rRNA depletion often is efficient\n",
    "* Accessing if the depletion step is successful through read distribution in genomic features\n",
    "    * How many reads map to ribosome RNA regions on the genome? \n",
    "    \n",
    "<blockquote>\n",
    "Which sample (S1-4) has the most insufficient rRNA depletion in lib prep? (refer to previous figure)\n",
    "</blockquote>\n",
    "\n",
    "#### 4.4 Confirmation of reduced/increased expression for knockdown/overexpressed genes in KO/OE samples\n",
    "* Is there a gene that is expected/known to be repressed or overexpressed?\n",
    "* Does the RNAseq result reflect expression change of this gene?\n",
    "* In our data, *PRDM11* expression is expected to be down since this is the gene knocked down in U2932 cells\n",
    "* NOT all exons are affected!\n",
    "* **Only those targeted by siRNA (small RNA interference) will be affected**\n",
    "\n",
    "<img src='ipynb_data/assets/Figure6.png', title = 'Figure6', width = 600, height = 600>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Quantify transcript abundance: featureCounts<a class=\"anchor\" id=\"06.4\"></a> <small>[[top](#top)]</small>\n",
    "* **Goal**\n",
    "    * Estimate number of reads mapped to gene features (e.g. gene, exon, etc.)\n",
    "* **Method**\n",
    "    * [featureCounts](http://bioinformatics.oxfordjournals.org/content/early/2012/10/25/bioinformatics.bts635), *version 2.5.2b*\n",
    "    * Ultrafast (~10 minutes for 50m reads), require low amount of memory (~4GB for human genome)\n",
    "    * Flexible options to count the reads based on specific mapping criteria or study purposes\n",
    "        * Gene-level or exon-level (for isoforms)\n",
    "        * Uniquely mapped reads\n",
    "        * Primary alignment\n",
    "        * Properly paired reads (if reads are paired-end)\n",
    "        * Mapping quality thresholds\n",
    "    * **Choose the option accordingly based on your experimental design!**\n",
    "* **Our data**\n",
    "    * Assigned\t55,910,832\n",
    "    * Unassigned_Ambiguity\t3,804,649\n",
    "    * Unassigned_NoFeatures\t4,192,188\n",
    "    * Unassigned_MappingQuality\t29,279,029\n",
    "\n",
    "<blockquote>\n",
    "```\n",
    "featureCounts  -s 0 -F GTF -t exon -g gene_id -Q 255 -J  --primary  -a $refgeneanno.gtf -T 2 -G $refgenome.fa -o $sample.star.featurecounts.raw_counts.txt $sample.star.merged.bam\n",
    "```\n",
    "</blockquote>\n",
    "\n",
    "<img src='ipynb_data/assets/Figure7.png', title = 'Figure7', width = 800, height = 580>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### BDS pipeline run report<a class=\"anchor\" id=\"06.12\"></a> <small>[[top](#top)]</small>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"1000\"\n",
       "            height=\"700\"\n",
       "            src=\"ipynb_data/assets/Run_RNAseq.bds.20161021_204005_581.report.html\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7f2ee82dd810>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import IFrame\n",
    "IFrame('ipynb_data/assets/Run_RNAseq.bds.20161021_204005_581.report.html', width=1000, height=700)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 5-8 are run by interactive R / Bioconductor practice (Notebook 02)\n",
    "\n",
    "### 5-7. Identify differentially expressed genes (DEGs) between conditions: DESeq2<a class=\"anchor\" id=\"06.5\"></a> <small>[[top](#top)]</small>\n",
    "* **Goal**\n",
    "    * Identify DEGs between groups of interest\n",
    "* **Method**\n",
    "    * [DESeq2](http://bioconductor.org/packages/release/bioc/html/DESeq2.html), *version 1.12.4*\n",
    "* **Our data**\n",
    "\n",
    "<img src='ipynb_data/assets/Figure8.png', title = 'Figure8', width = 700, height = 600>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 8. Identify biological processes and pathways enriched in genes of interest: clusterProfilter<a class=\"anchor\" id=\"06.8\"></a> <small>[[top](#top)]</small>\n",
    "* **Goal**\n",
    "    * Identify GO (gene ontology) and pathways significantly altered between groups of interest\n",
    "* **Method**\n",
    "    * [clusterProfiler](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html), *version 3.0.5*\n",
    "* **Our data**\n",
    "\n",
    "<img src='ipynb_data/assets/Figure9-1.png', title = 'Figure9-1', width = 800, height = 600>\n",
    "<img src='ipynb_data/assets/Figure9-2.png', title = 'Figure9-2', width = 800, height = 600>\n",
    "<img src='ipynb_data/assets/Figure9-3.png', title = 'Figure9-3', width = 800, height = 600>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Now ... let's move on to Notebook 02 (hands-on)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
